### PR TITLE
Better `math.cases` function

### LINKED ISSUE #1767: Revision of the cases function
### Description

I would love to see the `cases` function get a small revision.

To produce something like this in typst
grafik
you currently need to do something like this: 
`$ n |-> cases(3n+1 quad & 2 divides.not n, n/2  & 2 divides n) $`

I think it would be great if the `cases` function would use a syntax similar to the `mat` function.
I propose that creating the same output should instead work like this:
`$ n |-> cases(3n+1, 2 divides.not n; n/2, 2 divides n) $`
I want semicolons to declare the next line/entry and commas to separate the expression from the condition. All conditions should automatically be aligned.

This is similar to how LaTeX handles this (just better 🙃): `\cases{expression & condition \\  expression & condition}`

### Use Case

This would make using the cases function a lot more easy and enjoyable. Always doing `quad`/`"if"`/etc. and having to manually align is rather exhausting.

### PR DESCRIPTION
**Pull request by MDLC01**
Fixes #1767.

This PR renames the `math.cases` function to `math.group`, and adds a new `math.cases` function that lets you create case distinctions. This is more or less was was suggested in https://github.com/typst/typst/issues/1767#issuecomment-1657109071.

### Description

The new `math.cases` function accepts the same parameters as `math.group`, as well as an optional `separator` parameter. The elements of the variadic `cases` parameter go by pair: even elements correspond to values, while odd elements correspond to their conditions. There must be an even number of arguments.

```typ
$
  cases(
    x, "if" (x dot y)/2 <= 0,
    x + y, "if" x divides 2,
    x / y, "if" x in NN,
    0, "else",
  )
$
```
The cases have aligned conditions, that are separated from their corresponding values by commas.

```typ
#set math.cases(separator: [ ])
$
  abs(x) := cases(
    x, "if" x >= 0,
    -x, "if" x < 0,
  )
$
```
The cases have aligned conditions, without a visible separator between them and their corresponding values.

### Backward compatibility

This change is obviously backward incompatible, and will break a lot of documents. Although this is problematic, it must be noted that Typst is still in beta, and it is better making backward incompatible changes now than having to stick with poor decisions in the future.

With that said, any name idea that would allow us to keep `math.cases` for the current `math.cases` function (what this PR calls `math.group`) is welcome.

### Open questions

What should the default separator be? For now, I chose `$,quad$`. Other options include `$quad$`, `[ ]` and `[, ]`.

### CODE REVIEW

**Comment by T0mstone** [comment]
> With that said, any name idea that would allow us to keep math.cases for the current math.cases function (what this PR calls math.group) is welcome.

I strongly disagree. Given that it's also used for
- function definitions
- systems of equations
which have **nothing** to do with case distinctions.

Other than that, here is my honest reaction to reading this PR:
yes yes yes yes

**Comment by PgBiel** [comment]
As an initial opinion, I believe the separator should default to whitespace, as it is currently, as that'd be the most flexible and consistent option in respect to other math functions (where `,` means separating function arguments and not necessarily a literal comma, which can be inserted with `\,`). Also, would be nice to use the suggested `, ;` syntax similar to `mat` so that you can have arbitrarily many clauses (or even just 1 instead of 2).

**Comment by MDLC01** [comment]
I changed the default separator to a space and updated my original comment.

I'm not sure how useful the semicolon syntax would be, though. If you need to layout things like in a matrix, why not using `mat`? If `math.cases` is meant to be used for case distinctions, I don't think there is a need for more than two clauses.

**Comment by PgBiel** [comment]
> I changed the default separator to a space and updated my original comment.

Thanks.

> I'm not sure how useful the semicolon syntax would be, though. If you need to layout things like in a matrix, why not using `mat`? If `math.cases` is meant to be used for case distinctions, I don't think there is a need for more than two clauses.

Hmm, that's a fair point. I mostly just wanted the grouping to be more explicit (even if only allowing up to 2 clauses), i.e. to make it clear that `a, b;` are in the same line instead of using `a, b, c, d` which can lead to some confusion. But this is just an opinion, and I'd love to hear what others think about this.

**Comment by T0mstone** [comment]
>  using `a, b, c, d` which can lead to some confusion
We already have the same thing with `grid` and `table`, so I don't think it's that confusing.

**Comment by Eloitor** [comment]
If the `, ;` syntax is used, there is no need for two distinct functions `group` and `cases`...

**Comment by PgBiel** [comment]
> > using `a, b, c, d` which can lead to some confusion
> 
> We already have the same thing with `grid` and `table`, so I don't think it's that confusing.

True, great point.



> If the `, ;` syntax is used, there is no need for two distinct functions `group` and `cases`...

Also true! It could be a way to quickly specify columns without having to resort to a `grid`-like `columns` parameter.

**Comment by PgBiel** [comment]
Note that there's also the possibility of just staying with status quo and using alignment for different columns.

**Comment by MDLC01** [comment]
> Note that there's also the possibility of just staying with status quo and using alignment for different columns.

I think staying with the status quo (optionally renaming `math.cases` to `math.group`) is better than introducing a new function that does the same thing with a different syntax (`&` --> `,`, `,` --> `;`).

But adding a function that is made specifically for case distinction is the best idea imo.

**Comment by StrangeGirlMurph** [comment]
> Hmm, that's a fair point. I mostly just wanted the grouping to be more explicit (even if only allowing up to 2 clauses), i.e. to make it clear that `a, b;` are in the same line instead of using `a, b, c, d` which can lead to some confusion. But this is just an opinion, and I'd love to hear what others think about this.

I totally agree @PgBiel! I also prefer `a, b;`. Out of all the ideas (`a & b;`, `a & b \`, `a , b;`, `a , b`) I still like the initial one the most. Having that visual distinction between every line and following the already existing `mat` syntax just makes sense to me.

> If the `, ;` syntax is used, there is no need for two distinct functions `group` and `cases`...

Yes I think so too! `a , b;` is just very versatile. The separator option makes it perfect! You can easily group equations with `eqn1; eqn2; eqn3`, do the regular cases function with `a, x<3; b, x>= 3` and go beyond with something like `a,b,c;`. Now the questions is just what to call this function. I think `group` would be more accurate but `cases` might still be the better choice.

**Comment by MDLC01** [comment]
> Having that visual distinction between every line and following the already existing `mat` syntax just makes sense to me.

I agree that the absence of visual distinction is a drawback of the comma only syntax, but, as mentioned earlier by @T0mstone, this is already a problem with `grid` and `table`.

I have two problems with the `, ;` syntax. First, if `group` uses the same syntax as `mat` and does pretty much the same thing (displaying content in a grid), why not use `mat` instead (if necessary, we could allow its `delim` parameter to accept a dictionary with `left` and `right` keys)?[^1] Second, it makes the most common use of `group` (that is, simply having multiple rows grouped with a brace) awkward, as you now have to separate arguments with semicolons instead of commas.

[^1]: Although I guess `group` does something more: there is a `separator` property that does not exist on `mat`.

Basically, (as of the current state of this PR) we have four functions that are very similar:
- `mat` arranges math content in a grid (could be expressed as a `vec` with alignment points);
- `vec` arranges math content in a column (could be expressed as a `mat` with a single column);
- `group` arranges math content in a column, with a brace on the left (could be expressed as a `vec` with a brace on the left, and nothing on the right);
- `cases` arranges math content into two columns, with a brace on the left (could be expressed as a `vec` with a brace on the left and nothing on the right, using alignment points; or as a `mat` with two columns).

My reasoning is that we often want to layout the following things:
1. matrices,
2. vectors,
3. case distinctions,
4. rows of math text grouped by braces.

Each of those things should have a function that is specifically made for this purpose (in the current state of this PR, respectively, `mat`, `vec`, `cases` and `group`).

> If the `, ;` syntax is used, there is no need for two distinct functions `group` and `cases`...

There is no need for `vec` when `mat` is a thing either. Having a separate function is just more convenient.

An argument in favor of `, ;` that was not mentioned is that this syntax is "special" and cannot be recreated by the user. If we keep the PR as is, it will not be possible to recreate the `, ;` syntax. But if we instead implement `, ;`, users will be able to create their own, `cases` function that uses the comma only syntax.

TL;DR: No matter what syntax we choose, I think there should be four distinct functions in the end, each optimized for one of the four use cases mentioned above. We already have those for 1., 2. and 4.. We need to create one for 3.. Any other use case is uncommon enough that we should not sacrifice anything for it (but we can still try to make it not too annoying to do). I don't know which of the proposed syntax is best, although I prefer the current one.

**Comment by Eloitor** [comment]
@MDLC01, It is currently possible to use `, ;` in user defined functions:

https://typst.app/project/rhjKoozJQJV2NyVEXBtDh7

But it only works in math mode

**Comment by MDLC01** [comment]
I was not aware this was possible at all, thanks!

**Comment by laurmaedje** [comment]
I think another important point for having separate functions (even if they are all implemented on top of a common base function) is semantics. It helps both when reading the equations in source code and for semantic export (PDF, HTML) to know whether something is a vector, matrix, or case distinction.
